### PR TITLE
Bugfix/magetwo 61422 category max depth configuration

### PR DESCRIPTION
**Pull request by arnoudhgz**
### Description
The setting `Stores > Configuration > Catalog > Catalog > Category Top
Navigation > Maximal Depth` was actually never being used to limit the
topmenu.

Now the setting will be used as a condition for the category collection when it is fetched.

### Fixed Issues
1. magento/magento2#7543: Category Top Navigation / Maximal Depth configuration not working

### Manual testing scenarios
1. Change the setting in the backend
2. Refresh the invalidated cache
3. See that the setting actually is being used

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=12528) <br/>All committers have signed the CLA.

File: app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php
Comment (mzeis) [inline app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php:16]: Adding suppressions for CouplingBetweenObjects warnings is not allowed. The reason is that this increases Technical Debt. We understand this means more work, but refactoring this will help improve the code. Also, just adding an additional parameter would change the signature of the `@api` annotated class. Both changes mean that we probably should rebase this PR against `2.3-develop` so we can add breaking changes.

File: app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php
Comment (mzeis) [inline app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php:68]: Please use `private` visibility. The core team avoids `protected` for new properties (see [2.7](http://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html)). Also, new properties shouldn't have a name starting with an underscore (this is part of [PSR-2](http://www.php-fig.org/psr/psr-2/)).

File: app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php
Comment (mzeis) [inline app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php:98]: The rules for [backward compatible development](http://devdocs.magento.com/guides/v2.0/contributor-guide/backward-compatible-development/) only allow us to add a new argument to the constructor if it is optional (nullable) for changes to the same minor release. This means that adding the new parameter is okay if we re-base the PR against `2.3-develop` but for `2.2-develop` we would introduce breaking changes which developers are not expecting.

File: app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php
Comment (mzeis) [inline app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php:468]: Adding a public method to a class or an interface is another thing which we need to get approval for from a Magento architect as it changes the public signature of the `@api`-annotated class.

**Review by mzeis — CHANGES_REQUESTED** [review]
Hi @arnoudhgz,

thanks for your PR! You find a few comments from me inside the changelist. There are some parts that have to be refactored so they follow the Magento coding standards. It would be great if you can follow up on them.

**Comment by arnoudhgz** [comment]
@mzeis I accidentally closed the PR, I will create a new one and link it here

File: app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php
Comment (arnoudhgz) [reply app/code/Magento/Catalog/Model/ResourceModel/Category/Collection.php:16]: I will do a PR on 2.3-develop. 

The suppression is the same as the one from it's parent, so I'll keep that one.

**Comment by mzeis** [comment]
@arnoudhgz Thank you very much!

**Comment by arnoudhgz** [comment]
New PR on `2.3-develop`: https://github.com/magento/magento2/pull/12640

**Pull request closed without merging**
